mixer/pkg/runtime2/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,448 @@
+func (t *RoutingContext) DecRef() {
+	atomic.AddInt32(&t.refCount, -1)
+}
+
+// GetRefs returns the current reference count on the dispatch context.
+func (t *RoutingContext) GetRefs() int32 {
+	return atomic.LoadInt32(&t.refCount)
+}
+
+// New returns a new Dispatcher instance. The Dispatcher instance is initialized with an empty routing table.
+func New(identityAttribute string, handlerGP *pool.GoroutinePool, enableTracing bool) *Dispatcher {
+	return &Dispatcher{
+		identityAttribute: identityAttribute,
+		sessionPool:       newSessionPool(enableTracing),
+		statePool:         newDispatchStatePool(),
+		gp:                handlerGP,
+		context: &RoutingContext{
+			Routes: routing.Empty(),
+		},
+	}
+}
+
+// ChangeRoute changes the routing table on the Dispatcher which, in turn, ends up creating a new RoutingContext.
+func (d *Dispatcher) ChangeRoute(new *routing.Table) *RoutingContext {
+	newContext := &RoutingContext{
+		Routes: new,
+	}
+
+	d.contextLock.Lock()
+	old := d.context
+	d.context = newContext
+	d.contextLock.Unlock()
+
+	return old
+}
+
+// Check implementation of runtime.Dispatcher.
+func (d *Dispatcher) Check(ctx context.Context, bag attribute.Bag) (*adapter.CheckResult, error) {
+	s := d.beginSession(ctx, tpb.TEMPLATE_VARIETY_CHECK, bag)
+
+	var r *adapter.CheckResult
+	err := d.dispatch(s)
+	if err == nil {
+		r = s.checkResult
+		err = s.err
+	}
+
+	d.completeSession(s)
+
+	return r, err
+}
+
+// Report implementation of runtime.Dispatcher.
+func (d *Dispatcher) Report(ctx context.Context, bag attribute.Bag) error {
+	s := d.beginSession(ctx, tpb.TEMPLATE_VARIETY_REPORT, bag)
+
+	err := d.dispatch(s)
+	if err == nil {
+		err = s.err
+	}
+
+	d.completeSession(s)
+
+	return err
+}
+
+// Quota implementation of runtime.Dispatcher.
+func (d *Dispatcher) Quota(ctx context.Context, bag attribute.Bag, qma *runtime.QuotaMethodArgs) (*adapter.QuotaResult, error) {
+	s := d.beginSession(ctx, tpb.TEMPLATE_VARIETY_QUOTA, bag)
+	s.quotaMethodArgs = *qma
+
+	err := d.dispatch(s)
+	if err == nil {
+		err = s.err
+	}
+
+	r := s.quotaResult
+	d.completeSession(s)
+
+	return r, err
+}
+
+// Preprocess implementation of runtime.Dispatcher.
+func (d *Dispatcher) Preprocess(ctx context.Context, bag attribute.Bag, responseBag *attribute.MutableBag) error {
+	s := d.beginSession(ctx, tpb.TEMPLATE_VARIETY_ATTRIBUTE_GENERATOR, bag)
+	s.responseBag = responseBag
+
+	err := d.dispatch(s)
+	if err == nil {
+		err = s.err
+	}
+
+	d.completeSession(s)
+
+	return err
+}
+
+func (d *Dispatcher) dispatch(session *session) error {
+	// Capture the routing context locally. It can change underneath us. We also need to decRef before
+	// completing the call.
+	r := d.acquireRoutingContext()
+
+	// Lookup the value of the identity attribute, so that we can extract the namespace to use for route
+	// lookup.
+	identityAttributeValue, err := getIdentityAttributeValue(session.bag, d.identityAttribute)
```

Comment:
suggestion: move this up and don't acquire Routing context if identity attribute cannot be determined.

Reply:
Neat. Done.